settings: psa: add support for PS backend

dsseng opened the pull request:
- ~Depends on #94881 for tests and refactors not specific to PS, and contains it. Please review on per-commit basis~

- ~**tests: settings: add a persistence test (WIP)**~
- **settings: psa: rename from its**
- **settings: psa: add support for PS backend**
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/94681
- **tests: settings: cover PSA PS backend**

dsseng commented on tests/subsys/settings/functional/src/settings_basic_test.c:
```diff
@@ -657,3 +690,63 @@ ZTEST(settings_functional, test_direct_loading_filter)
+#define TEST_VALUE 0xfeedda7a
+ZTEST(settings_functional, test_persistence)
+{
+	// TODO: Z_TEST_SKIP_IFNDEF(CONFIG_TEST_SETTINGS_PERSISTENCE);
```
Here's a WIP test. It should probably have a separate test suite, but not sure. It is important for ITS and PS, since they write to storage using a delayed work, and are basically RAM-backed in runtime. So to make sure the PSA side of things actually works we need to reboot and validate it persists.

Could be included in the samples/subsys/settings as well I think, since that sample could showcase that after a reboot your settings are still there. It will also be used in the integration test, just as if in `tests/`

dsseng commented on subsys/settings/src/settings_psa.c:
```diff
@@ -6,59 +6,68 @@
 /* Ensure Key configured max size does not exceed reserved Key range */
-BUILD_ASSERT(sizeof(entries) / CONFIG_TFM_ITS_MAX_ASSET_SIZE <=
-	     ZEPHYR_PSA_SETTINGS_TFM_ITS_UID_RANGE_SIZE,
-	     "entries array exceeds reserved ITS UID range");
+BUILD_ASSERT(sizeof(entries) / SETTINGS_PSA_CHUNK_SIZE <=
+	     ZEPHYR_PSA_SETTINGS_PSA_UID_RANGE_SIZE,
+	     "entries array exceeds reserved PSA storage UID range");
```
Should I also add an assert for the case chunk number exceeds TF-M's storage entry count? That won't catch all misconfigurations, but will surely fail on ones where settings clearly won't fit even if no other ITS/PS entries are there

dsseng replied:
A `BUILD_ASSERT` should do the trick. It currently only validates if the array fits in the designated ID range, and I want to make it `min(range_size, max_entries_from_tfm_conf)`

seankyer commented on include/zephyr/psa/its_ids.h:
```diff
@@ -6,8 +6,8 @@
-/** UID range to be used by the TF-M ITS Settings backend. */
-#define ZEPHYR_PSA_SETTINGS_TFM_ITS_UID_RANGE_BEGIN 0x28000000
-#define ZEPHYR_PSA_SETTINGS_TFM_ITS_UID_RANGE_SIZE  0x10000 /* 64 Ki */
+/** UID range to be used by the PSA Settings backend. */
```
Might be showing my inexperience about the topic with this one, but is it possible that a system is configured to use both PS and ITS at the same time?

dsseng replied:
As I currently configured it it's a choice for either PS or ITS as the settings backend. iirc settings subsys only allows us to have one place to save entries to

seankyer commented on subsys/settings/src/settings_psa.c:
```diff
@@ -6,59 +6,68 @@
+#include <config_tfm.h>
+
+#define SETTINGS_PSA_CHUNK_SIZE PS_MAX_ASSET_SIZE
+#define SETTINGS_PSA_SET psa_ps_set
```
Nice, so I hadn't looked yet but the function signature between the `ps_set` and `its_set` are identical? Great.

seankyer commented on subsys/settings/Kconfig:
```diff
@@ -246,28 +249,48 @@ config SETTINGS_SHELL
+
+config SETTINGS_PSA_BACKEND_ITS
+	bool "PSA Internal Trusted Storage (ITS)"
+	depends on TFM_ITS_MAX_ASSET_SIZE != 0
```
Not sure if its required, but could add `depdends on TFM_PARTITION_INTERNAL_TRUSTED_STORAGE`

dsseng committed:
settings: tfm_psa: rename from its

Prepare for extending this backend to also allow storing settings in the
PS (Protected Storage).

Mostly file and kconfig renames, as well as updates
to comments and log messages.

Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

seankyer commented:
This is looking good. It makes sense to me and what I thought it should look like. I think you have some compliance things to deal with and if you could look at the issues raised by sonar that'd be great.

dsseng committed:
settings: tfm_psa: improve code style

Fix SonarQube warnings, move variable declarations to the
top of the block according to Zephyr code style

Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

tomi-font commented on subsys/settings/Kconfig:
```diff
@@ -143,17 +143,20 @@ config SETTINGS_CUSTOM
-	  Note: This settings backend compacts settings data into as few ITS nodes as possible.
-	  On every save, the entire settings array is written to ITS.
+	  This backend uses the PSA Internal Trusted Storage (ITS) service to store settings data.
+	  It can also be used for cases when settings need to be stored in a secure manner.
```
I'm not so convinced about this... if you wanna store data securely you would use a different API than settings in the first place

seankyer reviewed (dismissed):
Thank you for extending the ITS backend to PS.

tomi-font commented on subsys/settings/Kconfig:
```diff
@@ -265,28 +268,50 @@ config SETTINGS_SHELL
+if SETTINGS_TFM_PSA
+
+choice SETTINGS_TFM_PSA_BACKEND
+	prompt "PSA settings storage back-end"
```
```suggestion
	prompt "TF-M PSA Settings backend"
```

tomi-font commented on subsys/settings/src/settings_tfm_psa.c:
```diff
@@ -15,48 +15,67 @@
+#define SETTINGS_PSA_GET psa_its_get
+#else
+#error "No PSA backend selected"
+#endif
```
```suggestion
#endif /* CONFIG_SETTINGS_TFM_PSA_BACKEND */
```

tomi-font commented on subsys/settings/src/settings_tfm_psa.c:
```diff
@@ -15,48 +15,67 @@
+#define SETTINGS_PSA_CHUNK_SIZE PS_MAX_ASSET_SIZE
+#define SETTINGS_PSA_SET psa_ps_set
+#define SETTINGS_PSA_GET psa_ps_get
+#elif defined(CONFIG_SETTINGS_TFM_PSA_BACKEND_ITS)
```
Add some newline for readability?
```suggestion
#define SETTINGS_PSA_GET psa_ps_get

#elif defined(CONFIG_SETTINGS_TFM_PSA_BACKEND_ITS)
```

tomi-font commented on subsys/settings/src/settings_tfm_psa.c:
```diff
@@ -36,17 +36,32 @@ static const struct settings_store_itf settings_psa_itf = {
+#if defined(CONFIG_SETTINGS_TFM_PSA_BACKEND_PS)
+#include <psa/protected_storage.h>
+
+#define SETTINGS_PSA_CHUNK_SIZE PS_MAX_ASSET_SIZE
```
Why not `SETTINGS_PSA_MAX_ASSET_SIZE` or something like that?

tomi-font commented on subsys/settings/src/settings_tfm_psa.c:
```diff
@@ -157,17 +158,18 @@ static ssize_t settings_psa_read_fn(void *back_end, void *data, size_t len)
 	int ret;
 
-	for (int i = 0; i < entries_count; i++) {
+	for (i = 0; i < entries_count; i++) {
```
don't initialize `i` twice

tomi-font commented on subsys/settings/src/settings_tfm_psa.c:
```diff
@@ -260,21 +262,21 @@ void worker_persist_entries_struct_fn(struct k_work *work)
+	psa_status_t status = load_entries();
+
+	if (status != PSA_SUCCESS) {
+		/* If resource does not exist, we need to allocate it */
```
I think this comment should be moved down, right above `store_entries()`.

dsseng commented:
@tomi-font Thank you for review, I have addressed all the comments and refactored tests a bit

JarmouniA commented on subsys/settings/Kconfig:
```diff
@@ -143,17 +143,19 @@ config SETTINGS_CUSTOM
 
-config SETTINGS_TFM_ITS
-	bool "Internal Trusted Storage (ITS) settings backend"
+config SETTINGS_TFM_PSA
```
I believe renamed Kconfig symbols should be mentioned in the migration guide.

seankyer replied:
TFM settings backend is marked as experimental, so I don't think it's needed.

seankyer replied:
It is captured here https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html my interpretation is that we don't announce anything since it's still experimental. Not sure if we increment the version in the header however. @Kartben?

JarmouniA commented on samples/subsys/settings/sample.yaml:
Should be in a separate commit.

tomi-font commented on include/zephyr/psa/its_ids.h:
```diff
@@ -6,8 +6,8 @@
 #ifndef ZEPHYR_ITS_IDS_H_
 #define ZEPHYR_ITS_IDS_H_
```
Can you please fix this header guard to be `ZEPHYR_PSA_ITS_IDS_H_`?

tomi-font commented on include/zephyr/psa/its_ids.h:
No they don't, the current implementations we have isolate them, but the way we have been defining UID ranges so far has been so that they don't overlap even between PSA Crypto, ITS and PS, just in case and because the UID range is so huge that we can do this without downside so why not. Not super important.

dsseng committed:
settings: tfm_psa: add support for PS backend

Protected Storage is another type of secure (encrypted and
authenticated) storage available using PSA interfaces.

When targeting TF-M builds, allow making use of PS for storing settings

Fixes #94681

Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

dsseng committed:
doc: releases: migration-guide-4.4: add a note on SETTINGS_TFM_PSA

Add a note about a renamed Kconfig option

Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

dsseng committed:
tests: settings: cover PSA PS backend

Just like ITS, it can be tested on mps2/an521/cpu0/ns with TF-M.

Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

dsseng committed:
samples: settings: cover PSA PS backend

By default, run on mps2/an521/cpu0/ns with TF-M.

Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

tomi-font reviewed (dismissed):
One last remark. It would be good that this backend be covered in the maintainers file. Would you (and/or @seankyer) mind making a `file-groups` entry under the Settings area to cover all the files specific to the TF-M PSA backend and being a collaborator for it? You can add me as well. (Could be done as a separate PR or here, doesn't matter to me at least.)

dsseng commented:
force-pushed to rebase and fix merge conflict in migration guide

tomi-font commented on tests/subsys/settings/tfm_psa/testcase.yaml:
Oh, I see. Thanks GitHub...
